pkg/controllerhelpers/scylla.go

```diff
@@ -15,46 +16,110 @@ import (
-		return "", fmt.Errorf("service %s is of type %q instead of %q", naming.ObjRef(svc), svc.Spec.Type, corev1.ServiceTypeClusterIP)
+func GetScyllaHost(ctx context.Context, sc *scyllav1.ScyllaCluster, coreV1Client corev1client.CoreV1Interface, svc *corev1.Service) (string, error) {
+	if sc.Spec.ExposeOptions == nil || sc.Spec.ExposeOptions.BroadcastOptions == nil {
+		host, err := getServiceClusterIP(svc)
+		if err != nil {
+			return "", fmt.Errorf("can't get ClusterIP from service %q: %w", naming.ObjRef(svc), err)
+		}
+
+		return host, nil
+	}
+
+	switch sc.Spec.ExposeOptions.BroadcastOptions.Nodes.Type {
+	case scyllav1.BroadcastAddressTypeServiceClusterIP:
+		host, err := getServiceClusterIP(svc)
+		if err != nil {
+			return "", fmt.Errorf("can't get ClusterIP from service %q: %w", naming.ObjRef(svc), err)
+		}
+
+		return host, nil
+	case scyllav1.BroadcastAddressTypePodIP:
+		// Assume API's default
+		podIPSource := scyllav1.StatusPodIPSource
+		if sc.Spec.ExposeOptions.BroadcastOptions.Nodes.PodIP != nil {
+			podIPSource = sc.Spec.ExposeOptions.BroadcastOptions.Nodes.PodIP.Source
+		}
+		switch podIPSource {
+		case scyllav1.StatusPodIPSource:
+			pod, err := coreV1Client.Pods(sc.Namespace).Get(ctx, naming.PodNameFromService(svc), metav1.GetOptions{})
+			if err != nil {
+				return "", fmt.Errorf("can't get pod for service %q: %w", naming.ObjRef(svc), err)
+			}
+
+			host, err := getPodStatusPodIP(pod)
+			if err != nil {
+				return "", fmt.Errorf("can't get PodIP from pod %q status: %w", naming.ObjRef(pod), err)
+			}
+
+			return host, nil
+		default:
+			return "", fmt.Errorf("can't get PodIP from unsupported source: %q", podIPSource)
+		}
+	case scyllav1.BroadcastAddressTypeServiceLoadBalancerIngress:
+		host, err := getServiceExternalAddress(svc)
+		if err != nil {
+			return "", fmt.Errorf("can't get service %q external address: %w", naming.ObjRef(svc), err)
+		}
+
+		return host, nil
+	default:
+		return "", fmt.Errorf("unsupported broadcast address type: %q", sc.Spec.ExposeOptions.BroadcastOptions.Nodes.Type)
 	}
+}
 
+func getServiceClusterIP(svc *corev1.Service) (string, error) {
 	if svc.Spec.ClusterIP == corev1.ClusterIPNone {
-		return "", fmt.Errorf("service %s doesn't have a ClusterIP", naming.ObjRef(svc))
+		return "", fmt.Errorf("service does not have a ClusterIP")
 	}
 
 	return svc.Spec.ClusterIP, nil
 }
 
-func GetScyllaHost(statefulsetName string, ordinal int32, services map[string]*corev1.Service) (string, error) {
-	svcName := fmt.Sprintf("%s-%d", statefulsetName, ordinal)
-	svc, found := services[svcName]
-	if !found {
-		return "", fmt.Errorf("missing service %q", svcName)
+func getPodStatusPodIP(pod *corev1.Pod) (string, error) {
+	if len(pod.Status.PodIP) < 1 {
+		return "", fmt.Errorf("pod does not have a PodIP")
 	}
 
-	ip, err := GetScyllaIPFromService(svc)
-	if err != nil {
-		return "", err
+	return pod.Status.PodIP, nil
+}
+
+func getServiceExternalAddress(svc *corev1.Service) (string, error) {
```

Comment:
is this different from https://github.com/scylladb/scylla-operator/blob/eb144ac50175c040e486f47eb9468908181ce678/pkg/sidecar/identity/member.go#L55-L73

Reply:
not really, the two can be made into a shared function